### ISSUE TITLE
API executor subprocess broke. A service restart may be needed

### ISSUE THREAD

**Issue by Phandora** [wazuh/wazuh#16468]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.3.10-40323 | - | Manager | Packages | CentOS Linux 7 |

## Description

Hi team,

We have noticed that on some occasions the API stops working and the following errors appear in the `/var/ossec/logs/api.log` file:

<details>
<summary>API logs</summary>

```
20XX/XX/XX XX:XX:XX ERROR: API executor subprocess broke. A service restart may be needed
concurrent.futures.process._RemoteTraceback: 
'''
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/concurrent/futures/process.py", line 384, in wait_result_broken_or_wakeup
    result_item = result_reader.recv()
  File "/var/ossec/framework/python/lib/python3.9/multiprocessing/connection.py", line 256, in recv
    return _ForkingPickler.loads(buf.getbuffer())
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.10-py3.9.egg/wazuh/core/exception.py", line 594, in __init__
    error_details = self.ERRORS[self._code] if not cmd_error else extra_message
KeyError: 9011
'''

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.10-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 288, in execute_local_request
    data = await asyncio.wait_for(task, timeout=timeout)
  File "/var/ossec/framework/python/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.10-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 294, in execute_local_request
    raise exception.WazuhInternalError(901)
wazuh.core.exception.WazuhInternalError: Error 901 - API executor subprocess broke. A service restart may be needed


...


20XX/XX/XX XX:XX:XX ERROR: One of the API child processes terminated abruptly. The API process pool is not usable anymore. Please restart the Wazuh API
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.10-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 283, in execute_local_request
    task = loop.run_in_executor(pool, partial(self.run_local, self.f, self.f_kwargs,
  File "uvloop/loop.pyx", line 2695, in uvloop.loop.Loop.run_in_executor
  File "/var/ossec/framework/python/lib/python3.9/concurrent/futures/process.py", line 681, in submit
    raise BrokenProcessPool(self._broken)
concurrent.futures.process.BrokenProcessPool: A child process terminated abruptly, the process pool is not usable anymore

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.10-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 298, in execute_local_request
    raise exception.WazuhInternalError(900)




wazuh.core.exception.WazuhInternalError: Error 900 - One of the API child processes terminated abruptly. The API process pool is not usable anymore. Please restart the Wazuh API
```

</details>

This usually happens occasionally after executing the following commands to remove unused agents:

```
DELETE /agents?agents_list=all&status=disconnected&older_than=7d
DELETE /agents?agents_list=all&status=never_connected&older_than=7d
```

The Wazuh cluster includes `1 master and 11 workers` with the following CPU and memory limits:

- Master:
  - CPU: "7"
  - memory: 6Gi
- Workers:
  - CPU: "4.5"
  - memory: 3Gi

There are usually around 25,000 agents in total, with about 8,000 active agents on average.
Between 1,000 and 3,000 agents are removed daily.

After restarting the manager service everything seems to work properly.

____________________

### Notes

This issue also happened when the environment was in v4.3.6.

<details>
<summary>v4.3.6 API logs</summary>

```
20XX/XX/XX XX:XX:XX ERROR: API executor subprocess broke. A service restart may be needed
concurrent.futures.process._RemoteTraceback:
'''
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/concurrent/futures/process.py", line 384, in wait_result_broken_or_wakeup
    result_item = result_reader.recv()
  File "/var/ossec/framework/python/lib/python3.9/multiprocessing/connection.py", line 256, in recv
    return _ForkingPickler.loads(buf.getbuffer())
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.6-py3.9.egg/wazuh/core/exception.py", line 594, in __init__
    error_details = self.ERRORS[self._code] if not cmd_error else extra_message
KeyError: 9011
'''

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.6-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 288, in execute_local_request
    data = await asyncio.wait_for(task, timeout=timeout)
  File "/var/ossec/framework/python/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.6-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 294, in execute_local_request
    raise exception.WazuhInternalError(901)
wazuh.core.exception.WazuhInternalError: Error 901 - API executor subprocess broke. A service restart may be needed
```

</details>

Best regards,
Mayte Ariza

**Comment by mhamra** [comment]
### UPDATE

I've cloned branch v.4.3.11 and reviewed the exception source code line referred to in the log of the issue description. The error code 9011 doesn't exist in WazuhException.ERRORS dictionary, that's why the exception is being raised.

Searching instantiations of the Exception on framework code, I've found that wazuh.core.wazuh_socket.WazuhAsyncSocketJson.receive function instantiates an exception using the error number that comes in the response. It is probable that the exception raised comes from a socket with an error code 9011.

Reviewing the core source code, I've found that os_auth/local-server.c defines an ERRORS array with an error 9011

```
static const struct {
    int code;
    char *message;
} ERRORS[] = {
    { 9001, "Internal error" },
    { 9002, "Parsing JSON input" },
    { 9003, "No such function" },
    { 9004, "No such argument" },
    { 9005, "No such name" },
    { 9006, "No such IP" },
    { 9007, "Duplicate IP" },
    { 9008, "Duplicate name" },
    { 9009, "Issue generating key" },
    { 9010, "No such agent ID" },
    { 9011, "Agent ID not found" },
    { 9012, "Duplicate ID" },
    { 9013, "Maximum number of agents reached" },
    { 9014, "Invalid Group(s) Name(s)" },
    { 9015, "Cannot execute this request on a worker node" }
};

```

WazuException should manage non-existent error codes to avoid future problems with errors not defined in the ERRORS dictionary. This is how the code is implemented in 4.3.11 and it is still present in the master branch.
```
        error_details = self.ERRORS[self._code] if not cmd_error else extra_message
        if isinstance(error_details, dict):
            code_message, code_remediation = error_details.get('message', ''), error_details.get('remediation', None)
        else:
            code_message, code_remediation = error_details, None

        if not cmd_error:
            if extra_message:
                if isinstance(extra_message, dict):
                    self._message = code_message.format(**extra_message)
                else:
                    self._message = "{0}: {1}".format(code_message, extra_message)
            else:
                self._message = code_message
        else:
            self._message = extra_message
```
the proposed change is the following
```
        if not cmd_error and self._code in self.ERRORS:
            error_details = self.ERRORS[self._code] if not cmd_error else extra_message
            if isinstance(error_details, dict):
                code_message, code_remediation = error_details.get('message', ''), error_details.get('remediation', None)
            else:
                code_message, code_remediation = error_details, None
            if extra_message:
                if isinstance(extra_message, dict):
                    self._message = code_message.format(**extra_message)
                else:
                    self._message = "{0}: {1}".format(code_message, extra_message)
            else:
                self._message = code_message
        else:
            self._message = extra_message
```

**Comment by mhamra** [comment]
## Checks

The following elements have been updated or reviewed (should also be checked if no modification is required):
- [x] Tests (unit tests, API integration tests).
